Resources/ProtobufDumper/ProtobufDumper/ExecutableScanner.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.IO;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace ProtobufDumper
+{
+    class ExecutableScanner
+    {
+        private static readonly Regex ProtoFileNameRegex = new Regex( @"^[a-zA-Z_0-9\\/.]+\.proto$", RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.ExplicitCapture );
+
+        public delegate bool ProcessCandidate( string name, byte[] buffer );
+
+        public static void ScanFile( string fileName, ProcessCandidate processCandidate )
+        {
+            Console.WriteLine( "Loading binary '{0}'...", fileName );
+
+            ScanFile( File.ReadAllBytes( fileName ), processCandidate );
+        }
+
+        private static void ScanFile( byte[] data, ProcessCandidate processCandidate )
+        {
+            const char markerStart = '\n';
+            const int markerLength = 2;
+
+            var scanSkipNull = 0;
+            var scanSkipTotal = 0;
+
+            for ( var i = 0; i < data.Length - 1; i++ )
+            {
+                var currentByte = data[ i ];
+                var expectedLength = data[ i + 1 ];
+
+                if ( currentByte != markerStart ) continue;
+
+                var y = i;
+                for ( ; y < data.Length; y++ )
+                {
+                    if ( data[ y ] != 0 ) continue;
+
+                    if ( scanSkipNull == 0 ) break;
+                    scanSkipNull -= 1;
+                }
+
+                if ( y == data.Length ) continue;
+
+                var length = y - i;
+
+                if ( length < markerLength || length - 2 < expectedLength ) continue;
+
+
+                var bufferSpan = new ReadOnlySpan<byte>( data, i, length );
+                var nameSpan = bufferSpan.Slice( 2, expectedLength );
+
+                var protoName = Encoding.ASCII.GetString( nameSpan );
+
+                if ( !ProtoFileNameRegex.IsMatch( protoName ) ) continue;
+
+                if ( !processCandidate( protoName, bufferSpan.ToArray() ) )
```

Comment:
Can we keep this as a `Span`, or rent an array from an `ArrayPool`?

Reply:
I can keep this as a Span, but ultimately protobuf-net needs a Stream

Reply:
This stays as a Span until it needs to be deserialized, where it is copied into a MemoryStream. I don't know what options we really have.